core/src/main/scala/kafka/server/ClientMetricsManager.scala

```diff
@@ -0,0 +1,93 @@
+    compressionTypes.toList
+  }
+
+  def processGetSubscriptionRequest(request: RequestChannel.Request,
+                                    throttleMs: Int): GetTelemetrySubscriptionResponse = {
+    val subscriptionRequest = request.body[GetTelemetrySubscriptionRequest]
+    val clientInfo = CmClientInformation(request, subscriptionRequest.getClientInstanceId.toString)
+    _instance.processGetSubscriptionRequest(subscriptionRequest, clientInfo, throttleMs)
+  }
+
+}
+
+class ClientMetricsManager {
+  val clientInstanceCache = ClientMetricsCache.getInstance
+  def getCacheSize = clientInstanceCache.getSize
+  def clearCache() = clientInstanceCache.clear()
+  def getClientInstance(id: Uuid) = clientInstanceCache.get(id)
+
+  def processGetSubscriptionRequest(subscriptionRequest: GetTelemetrySubscriptionRequest,
+                                    clientInfo: CmClientInformation,
+                                    throttleMs: Int): GetTelemetrySubscriptionResponse = {
+    var clientInstanceId = subscriptionRequest.getClientInstanceId
+    if (clientInstanceId == null || clientInstanceId == Uuid.ZERO_UUID) {
+      clientInstanceId = Uuid.randomUuid()
+    }
+    var clientInstance = getClientInstance(clientInstanceId)
+    if (clientInstance == null) {
+      clientInstance = createClientInstance(clientInstanceId, clientInfo)
+    }
+
+    val data =  new GetTelemetrySubscriptionsResponseData()
+        .setThrottleTimeMs(throttleMs)
+        .setClientInstanceId(clientInstanceId)
+        .setSubscriptionId(clientInstance.getSubscriptionId) // TODO: should we use LONG instead of int?
+        .setAcceptedCompressionTypes(getSupportedCompressionTypes.asJava)
+        .setPushIntervalMs(clientInstance.getPushIntervalMs)
+        .setDeltaTemporality(true)
+        .setErrorCode(Errors.NONE.code())
+        .setRequestedMetrics(clientInstance.getMetrics.asJava)
+
+    if (clientInstance.isDisabledForMetricsCollection) {
+      info(s"Metrics collection is disabled for the client: ${clientInstance.getId.toString}")
+      data.setErrorCode(Errors.INVALID_CONFIG.code)
+    }
+
+    new GetTelemetrySubscriptionResponse(data)
+  }
+
+  def updateSubscription(groupId :String, properties :Properties) = {
+    ClientMetricsConfig.updateClientSubscription(groupId, properties, clientInstanceCache)
+  }
+
+  def createClientInstance(clientInstanceId: Uuid, clientInfo: CmClientInformation): CmClientInstanceState = {
+    val clientInstance = CmClientInstanceState(clientInstanceId, clientInfo,
+                                               ClientMetricsConfig.getClientSubscriptionGroups)
+
+    // Add to the cache and if cache size > max entries then time to make some room by running
+    // GC to clean up all the expired entries in the cache.
+    clientInstanceCache.add(clientInstance)
+    if (clientInstanceCache.getSize >  CM_CACHE_MAX_SIZE) {
+      ClientMetricsCache.runGC()
```

Comment:
I think it is better to not do this per request, you could DoS the broker by creating a lot of new unique Client cache instances that eventually exceeds the MAX_SIZE. This would result in a O(n) cache GC scan on each such request, and since all the TTLs would be in the future nothing would be GC:ed.

So, I recommend perhaps expediting the standard cache GC timer (to 1s or so)

Reply:
But even in that case, GC would be triggered only after CM_CACHE_GC_INTERVAL but I liked your idea of expiding the cache b bumping up the GC timer.. let me revisit the cache code